### ISSUE TITLE
Add more properties to quickfix/location list items -- to aid highlighting full text of interest of items

### ISSUE THREAD

**Issue by poetaman** [vim/vim#8003]
**Is your feature request about something that is currently impossible or hard to do? Please describe the problem.**
It is difficult to highlight the entire text pointed by an item in quickfix/location list. Currently quickfix/location list items have property ``lnum`` & ``col`` which point to the start position of item's text, but does not contain information it's end position.

**Describe the solution you'd like**
Be able to highlight entire text of quickfix/locationlist items. For this add:

2 properties to quickfix/location list items:
* ``end_lnum``: line number of end of text in item
* ``end_col``: column number of end of text in item

Optionally also add 1 property to {what} dictionary:
``pattern``: This is the pattern that applies to all quickfix/location list items. Store the pattern used by the user in commands like ``:vimgrep`` & ``:grep`` in this. Currently there is ``pattern`` property in items, but it would be too much replication for such commands (as pattern is essentially the same).

**Describe alternatives you've considered**
Do a regular expression match starting from the ``lnum`` & ``col`` of quickfix/location list item.

**Comment by bfrg** [comment]
First of all, I am not against this feature request but in my opinion there are some issues that need to be addressed first.

> Optionally also add 1 property to `{what}` dictionary:
`pattern`: This is the pattern that applies to all quickfix/location list items. Store the pattern used by the user in commands like `:vimgrep` & `:grep` in this. Currently there is `pattern` property in items, but it would be too much replication for such commands (as pattern is essentially the same).

What do you do when someone runs `:vimgrepadd` or `:grepadd` with a different search pattern? You probably have to store a list of patterns. But then you also need to store the indexes of the quickfix items (or a range) each pattern applies to. Alternatively, add an additional quickfix item `regex` for each quickfix item (since `pattern` is already used in a different context).

> Store the pattern used by the user in commands like `:vimgrep` & `:grep` in this.

For `:vimgrep` this is probably doable, but there is no easy way Vim can extract the search pattern of a `:grep` command since the command relies on `grepprg` which can be set to _anything_. Maybe a new option `grepcmdexpr` could be added that users can set to a function that takes the full grep command as argument and returns the pattern?

Even if the `:grep` pattern could be extracted somehow, keep in mind that it is not directly usable in Vim since Vim has its own regex flavor. Depending on `grepprg` the command could be using BRE (basic regular expressions), ERE (extended regular expressions), PCRE (Perl-compatible regular expressions), or anything else. In order to use the pattern for highlighting you would have to convert it to Vim regular expressions first.

I like the idea of adding the properties `end_line` and `end_col`. But similar to `pattern`, Vim can set those entries only for `:vimgrep`. For `:grep` we would have to tell Vim how to parse the output of `grepprg` to obtain these values, for example by introducing a new option `grepoutexpr`. Users could then set it to a function that takes the output of `:grep` as a list of strings and returns a list of quickfix entries with `end_line` and `end_col` entries. AFAIK there are not many grep programs that provide something like `end_line` and `end_col`. `grep(1)` doesn't even output the column number of a matched line. `git-grep`, `the_silver_searcher` and `ripgrep` have the `--column` option but apart from `ripgrep` none of them print the length of the match. `ripgrep --json` is the only way to obtain the range of a match. With a new option `grepoutexpr` ripgrep's json output could be converted to a list of quickfix entries with `end_line` and `end_col` set. See following [link](https://docs.rs/grep-printer/0.1.5/grep_printer/struct.JSON.html#object-submatch) for the exact json format.

If a new option like `grepoutexpr` is added, maybe we should also consider adding `makeoutexpr` for the `:make` command which will be used instead of `errorformat` to obtain a quickfix list with `end_line` and `end_col` entries. Compilers that provide a range for each error are, for example, gcc with `-fdiagnostics-format=json`, clang with `-fdiagnostics-print-source-range-info`, or rustc with the `--error-format=json` flag.

Now, even if all the above new options are added and users manage to set them appropriately so that `end_line` and `end_col` are obtained from search commands, there's still another issue regarding highlighting the matches. I'm assuming this would be done using text-properties. The nice thing about text-properties is that they move with the text. If you delete a line containing a text-property, the highlighting will disappear and reappear after `:undo`. But quickfix items are (unfortunately) not updated like this. If you check the quickfix item for a particular line that you first delete, and then `:undo`, the item's `lnum` has increased by one. Hence, after making some changes to the buffer the text highlightings and quickfix items won't overlap anymore. This will look like a bug and probably annoy some users. I don't know if this can be improved on the Vim side, or if it even makes sense because of performance reasons.

Like I said, I'm not against any of this but it's not as straight forward as one might think to make everything work together without any hiccups.

**Comment by lacygoill** [comment]
Possibly fixed by the patch [8.2.3019](https://github.com/vim/vim/releases/tag/v8.2.3019) which adds the `end_lnum` and `end_col` properties to each entry of a quickfix list.

Usage example:
```vim
vim9script

prop_type_add('QuickFixHighlight', {highlight: 'QuickFixLine'})

augroup QuickFixHighlight
    autocmd!
    autocmd QuickFixCmdPost * QuickFixHighlight()
augroup END

def QuickFixHighlight()
    for entry in getqflist()
        if entry.bufnr == 0
            continue
        endif
        entry.bufnr->bufload()
        prop_add(entry.lnum, entry.col, {
            type: 'QuickFixHighlight',
            bufnr: entry.bufnr,
            end_lnum: entry.end_lnum == 0 ? entry.lnum : entry.end_lnum,
            end_col: entry.end_col == 0 ? entry.col : entry.end_col,
        })
    endfor
enddef

helpgrep foobar
```
